// slimcore/lib/slimcore.js
/// <reference path="slimcore.d.ts" />
"use strict";
const version = require("./version");
const native = module.require('../bin/slimcore.node');
// version
exports.getVersion = version.getVersion;
exports.getApiVersion = version.getApiVersion;
// slimcore
exports.createSlimCoreInstance = native.createSlimCoreInstance;
exports.createVideoBindingRenderer = native.createVideoBindingRenderer;
exports.createVideoBindingScreenShare = native.createVideoBindingScreenShare;
exports.createPepperFrameSink = native.createPepperFrameSink;
exports.createCanvasFrameSink = native.createCanvasFrameSink;
exports.createDataSource = native.createDataSource;
exports.createDataSink = native.createDataSink;
exports.queryDeviceRotation = native.queryDeviceRotation;
exports.SlimCore = native.SlimCore;
exports.CallInterface = native.CallInterface;
exports.ScreenScraper = native.ScreenScraper;
exports.VideoBinding = native.VideoBinding;
exports.VideoBindingRenderer = native.VideoBindingRenderer;
exports.VideoBindingScreenShare = native.VideoBindingScreenShare;
exports.FrameSink = native.FrameSink;
exports.CanvasFrameSink = native.CanvasFrameSink;
exports.PepperFrameSink = native.PepperFrameSink;
exports.ContentSharing = native.ContentSharing;
exports.DataChannel = native.DataChannel;
exports.DataSource = native.DataSource;
exports.DataSink = native.DataSink;
exports.Trouter = native.Trouter;
exports.TrouterListener = native.TrouterListener;
exports.TrouterRequest = native.TrouterRequest;
exports.TrouterResponse = native.TrouterResponse;
exports.Enums = native.Enums;
