// main-db-exporter/dist/batch-index.js
"use strict";
const fs = require('fs');
const path = require('path');
const DOC_HEAD = `<!DOCTYPE html>
<html>
    <head>
        <title>Archived conversations</title>
        <meta charset="utf-8">
    </head>
<body>
    <div class="header">
        <h1 class="conversations">Archived conversations</h1>
        <ul class="exported-accounts" id="accounts">
\n`;
class BatchIndex {
    constructor(outPath) {
        this.exportedItems = [];
        this.outPath = outPath;
    }
    addEntry(account, indexPath) {
        this.exportedItems.push({
            account,
            indexPath: path.join(indexPath, 'index.html')
        });
    }
    write() {
        if (!this.exportedItems.length) {
            return;
        }
        this.stream = fs.createWriteStream(path.join(this.outPath, 'index.html'));
        this.stream.write(DOC_HEAD);
        this.exportedItems.forEach((exportItem) => {
            this.writeOne(exportItem);
        });
        this.stream.end(`
        </ul>
    </body>
</html>`);
    }
    writeOne(exportItem) {
        let relativePath = exportItem.indexPath.startsWith(this.outPath) ? exportItem.indexPath.replace(this.outPath, '.') : exportItem.indexPath;
        relativePath = relativePath.replace('#', '%23');
        const template = `<li class="account">
        <a href="${relativePath}">${exportItem.account.fullName} (${exportItem.account.skypeName})</a>
        </li>\n`;
        this.stream.write(template);
    }
}
exports.BatchIndex = BatchIndex;
