// main-db-exporter/dist/html-formatter.js
"use strict";
const fs = require('fs');
const path = require('path');
const exporter_types_1 = require('./exporter-types');
const file_output_1 = require('./file-output');
const html_index_1 = require('./html-index');
const msg_body_transform_1 = require('./msg-body-transform');
class HtmlFormatter {
    fileExtension() {
        return 'html';
    }
    formatMessage(row) {
        if (!row.body_xml) {
            return null;
        }
        const translatedType = exporter_types_1.MESSAGE_TYPE_MAP[row.type];
        if (translatedType) {
            let timeString = '';
            if (row.timestamp) {
                timeString = msg_body_transform_1.formatTimestamp(row.timestamp);
            }
            let body = '';
            if (row.body_xml) {
                body = msg_body_transform_1.sanitizeXml(row.body_xml, row.id);
                if (body === '') {
                    console.log(`Empty transformed message, id:${row.id} body: ${row.body_xml}`);
                }
            }
            let fromDisplayName = msg_body_transform_1.sanitizeXml(row.from_dispname);
            const result = `<li class="message" id="${row.id}" >
<div>
<span class="author">${fromDisplayName}</span>
<span class="timestamp">${timeString}</span>
</div>
<div class="message-body">${body}</div>
</li>`;
            return result;
        }
        else {
            return null;
        }
    }
    formatConversation(row) {
        if (!row || !row.id || !row.identity || !row.type) {
            return null;
        }
        const blocked = !!row.is_blocked;
        let str = '';
        if (blocked) {
            str += '<div class="blocked">blocked converation</div>';
        }
        return str;
    }
    fileHeader(row) {
        return `<!DOCTYPE html>
<html>
<head>
<title>${row.displayname}</title>
<meta charset="utf-8" />
<link rel="stylesheet" href="style.css">
</head>
<body class="conversation">
<script>
document.addEventListener("DOMContentLoaded", function(event) {
        var e = document.body;
        e.scrollTop = e.scrollHeight - e.clientHeight;
});
</script>
`;
    }
    fileFooter() {
        return '</body>\n</html>\n';
    }
    messageSeparatorHeader() {
        return '\n<ul class="messages" id ="messages">\n';
    }
    messageSeparatorFooter() {
        return '</ul>\n';
    }
    messageSeparator() {
        return '\n';
    }
    createOutput(completeOutPath, conversation) {
        return new file_output_1.FileOutput(this, completeOutPath, conversation);
    }
    init(outPath) {
        let sourcePath = path.join(__dirname, 'resources');
        this.copyFile(sourcePath, outPath, 'style.css');
        this.copyFile(sourcePath, outPath, 'skype.svg');
        this.copyFile(sourcePath, outPath, 'default1.png');
        this.copyFile(sourcePath, outPath, 'default2.png');
        this.index = new html_index_1.HtmlIndex(outPath);
    }
    getIndex() {
        return this.index;
    }
    copyFile(sourcePath, outPath, fileName) {
        const rs = fs.createReadStream(path.join(sourcePath, fileName));
        rs.on('error', (err) => console.log(`Error while trying to copy: ${fileName}: ${err}`));
        const ws = fs.createWriteStream(path.join(outPath, fileName));
        ws.on('error', (err) => console.log(`Error while trying to copy: ${fileName}: ${err}`));
        rs.pipe(ws);
    }
}
exports.HtmlFormatter = HtmlFormatter;
