// main-db-exporter/dist/html-index.js
"use strict";
const fs = require('fs');
const path = require('path');
const msg_body_transform_1 = require('./msg-body-transform');
const DOC_HEAD = `<!DOCTYPE html>
<html>
    <head>
        <title>Archived conversations</title>
        <meta charset="utf-8">
        <link rel="stylesheet" href="style.css">
    </head>
    <body>
        <script>
            document.addEventListener("DOMContentLoaded", function(event) {
                function selectConversation(element) {
                    if (selectedConvLi) {
                        selectedConvLi.className = "conversations";
                    }
                    var href = element.getAttribute('href');
                    selectedConversationHeader.innerHTML = element.innerHTML;
                    selectedConversationDiv.innerHTML='<object class="object" type="text/html" data="' + href + '" ></object>';
                    selectedConvLi = element.parentNode;
                    selectedConvLi.className = "conversations-sel"
                }
                function convCallback(event) {
                    if (event.target.tagName !== 'A') {
                        return;
                    }
                    selectConversation(event.target);
                    event.preventDefault();
                }

                var selectedConvLi = null;
                var selectedConversationDiv = document.getElementById("selected-conversation-placeholder");
                var selectedConversationHeader = document.getElementById("selected-conversation-header");
                var conversations = document.getElementById('conversations');
                conversations.addEventListener('click', convCallback, false);
                selectConversation(conversations.firstElementChild.firstElementChild);
            });
        </script>
        <div class="left">
            <div class="header">
                <h1 class="conversations">Archived conversations</h1>
\n`;
class HtmlIndex {
    constructor(outPath) {
        this.conversations = [];
        this.stream = fs.createWriteStream(path.join(outPath, 'index.html'));
        this.stream.write(DOC_HEAD);
    }
    addEntry(conversation) {
        this.conversations.push(conversation);
    }
    close(accountInfo) {
        this.writeHeader(accountInfo);
        const sorted = this.conversations.sort((a, b) => { return b.maxTimestamp - a.maxTimestamp; });
        sorted.forEach((conv) => { this.writeOne(conv); });
        this.stream.end(`
        </ul>
        </div>
        <div id= "selected-conversation" class="right" >
            <h1 class="conversation" id="selected-conversation-header"></h1>
            <div id= "selected-conversation-placeholder">
            </div>
        </div>
    </body>
</html>`);
    }
    writeHeader(accountInfo) {
        const time = this.formatTimestamp(new Date());
        const fullName = msg_body_transform_1.sanitizeXml(accountInfo.fullName, '');
        let header = `<table>
                <tbody>
                    <tr><th>User</th><td id="hdr-user">${fullName} <small>(${accountInfo.skypeName})</small></td></tr>
                    <tr><th>Exported</th><td id="hdr-exported">${time}</td></tr>
                    <tr><th>Total</th><td id="hdr-stats">${this.conversations.length} conversations</td></tr>
                </tbody>
            </table>
        </div>
        <ul class="conversations" id="conversations">
        `;
        this.stream.write(header);
    }
    writeOne(conversation) {
        const timestamp = this.formatTimestamp(new Date(conversation.maxTimestamp * 1000));
        const displayName = msg_body_transform_1.sanitizeXml(conversation.displayname, '');
        const template = `
        <li class="conversations">
            <a href="${conversation.fileName}#end" target="selected-chat">${displayName}</a>
            <div>
                <span class ="messageCount">${conversation.messageCount} ${this.formatMessageCount(conversation.messageCount)}</span>
                <span class ="timestamp-conv">Last from: ${timestamp}</span>
            </div>
        </li>\n`;
        this.stream.write(template);
    }
    formatTimestamp(dt) {
        return `${dt.toLocaleDateString()} ${dt.toLocaleTimeString()}`;
    }
    formatMessageCount(messageCount) {
        return messageCount > 1 ? 'messages' : 'message';
    }
}
exports.HtmlIndex = HtmlIndex;
